### ISSUE TITLE
whether is it better to add class_weight to validation dataset or not?

### ISSUE THREAD

**Issue by loveychen** [tensorflow/tensorflow#42647]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.x
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

## set `class_weight` to validation dataset

It is common for us to set `class_weight` to something like inverse of class supports  to `Model.fit()` when dealing with imbalance classification problems. And as the code implemented in `tensorflow.python.keras.engine.data_adapter._make_class_weight_map_fn`, `class_weight` will eventually be converted to `sample_weight`.

However, the current implementation of `Model.fit()` only add `class_weight` to the training dataset but not to the validation dataset

```python
# for training dataset
data_handler = data_adapter.DataHandler(
  x=x,
  y=y,
  sample_weight=sample_weight,
  batch_size=batch_size,
  steps_per_epoch=steps_per_epoch,
  initial_epoch=initial_epoch,
  epochs=epochs,
  shuffle=shuffle,
  class_weight=class_weight,
  max_queue_size=max_queue_size,
  workers=workers,
  use_multiprocessing=use_multiprocessing,
  model=self,
  steps_per_execution=self._steps_per_execution)

# for validation dataset
self._eval_data_handler = data_adapter.DataHandler(
    x=val_x,
    y=val_y,
    sample_weight=val_sample_weight,
    batch_size=validation_batch_size or batch_size,
    steps_per_epoch=validation_steps,
    initial_epoch=0,
    epochs=1,
    max_queue_size=max_queue_size,
    workers=workers,
    use_multiprocessing=use_multiprocessing,
    model=self,
    steps_per_execution=self._steps_per_execution)
```

which will output training loss and validation loss with (maybe) large gap, such as:

```
Epoch 2/10000
128/128 [==============================] - 63s 489ms/step - loss: 379.6439 - acc: 0.0073 - acc_f1: 0.0077 - cm: 10485.7598 - val_loss: 4.9299 - val_acc: 0.0076 - val_acc_f1: 0.0081 - val_cm: 2621.4399
128/128 [==============================] - 63s 492ms/step - loss: 374.9651 - acc: 0.0078 - acc_f1: 0.0099 - cm: 10485.7598 - val_loss: 4.8990 - val_acc: 0.0078 - val_acc_f1: 0.0120 - val_cm: 2621.4399
```

**I am wondering whether it is a good practice to add the same `class_weight` parameter to the training dataset and the validation dataset, in order to keep training loss and validation loss with the same scale**.

## normalized loss by the sum of `sample_weight`

And by the way, while reducing the loss by the size of a batch, the current loss is normalized by the size of `weighted_loss` (ref. to `tensorflow.python.keras.utils.losses_utils.reduce_weighted_loss`),  **whether is it better to normalize the loss by the sum of `sample_weight`**?

**Will this change the current api? How?**
Almost no changes.

For adding `class_weight` to validation dataset, only add `class_weight=class_weight` when calling `data_adapter.DataHandler`

```python
self._eval_data_handler = data_adapter.DataHandler(
    x=val_x,
    y=val_y,
    sample_weight=val_sample_weight,
    batch_size=validation_batch_size or batch_size,
    steps_per_epoch=validation_steps,
    initial_epoch=0,
    epochs=1,
    shuffle=shuffle,
    class_weight=class_weight,
    max_queue_size=max_queue_size,
    workers=workers,
    use_multiprocessing=use_multiprocessing,
    model=self,
    steps_per_execution=self._steps_per_execution)
```

For normalizing loss with the sum of `sample_weight`,  only changes the code snippet of `python.keras.utils.losses_utils.compute_weighted_loss`

from 

```python
    ....
    # Apply reduction function to the individual weighted losses.
    loss = reduce_weighted_loss(weighted_losses, reduction)
    # Convert the result back to the input type.
    loss = math_ops.cast(loss, input_dtype)
    return loss
```
to 

```python
   
    if sample_weight is None:
      sample_weight = array_ops.ones_like(losses, dtype=losses.dtype)
   ...
    # Apply reduction function to the individual weighted losses
    if reduction == ReductionV2.NONE:
      loss = weighted_losses
    else:
      loss = math_ops.reduce_sum(weighted_losses)
      if reduction == ReductionV2.SUM_OVER_BATCH_SIZE:
        loss = _safe_mean(loss, math_ops.reduce_sum(sample_weight))
    
    # Convert the result back to the input type.
    loss = math_ops.cast(loss, input_dtype)
    return loss
```

**Who will benefit with this feature?**
everyone if it is a good practice.

**Any Other info.**

**Comment by gowthamkpr** [comment]
You are not supposed to apply weights to your validation set since it is supposed to measure your model's performance. If you'll do that you'll probably get better results for validation but once your model is deployed it will perform worse on new data. Weighting, resampling techniques etc. - they all should be done on the training set only!

**Comment by loveychen** [comment]
thanks, @gowthamkpr. You are right, the loss gap is not that important, and we should keep our eyes on the performance on new data.
